File: php/muse/util/Sequences.php
<?php
    namespace muse\util\Sequences;

    require_once "muse/algorithms/sorting/QuickSort.php";
    require_once "muse/util/RandomFactory.php";

    use muse\algorithms\sorting\QuickSort;
    use muse\util\RandomFactory;

    function inspect(array &$arr): void {
        print("[\n");
        foreach ($arr as $i => $v) {
            printf("\t#%04X  -->  %d\n", $i, $v);
        }
        print("]\n");
    }

    function isSorted(array &$arr): bool {
        for ($i = 1, $length = count($arr); $i < $length; $i++) {
            if ($arr[$i] < $arr[$i - 1]) {
                return false;
            }
        }

        return true;
    }

    function parityChecksum(array &$arr): int {
        $checksum = 0;

        foreach ($arr as $v) {
            $checksum ^= $v;
        }

        return $checksum;
    }

    function reverse(array &$arr): void {
        $k;
        $temp;

        for ($i = 0, $bound = count($arr) >> 1, $length = count($arr); $i < $bound; $i++) {
            $k = $length - $i - 1;
            $temp = $arr[$i];
            $arr[$i] = $arr[$k];
            $arr[$k] = $temp;
        }
    }

    function shuffle(array &$arr): void {
        $k;
        $temp;

        RandomFactory\seed();
        for ($i = 0, $length = count($arr); $i < $length; $i++) {
            $k = RandomFactory\genIntN($i, $length);
            $temp = $arr[$i];
            $arr[$i] = $k;
            $k = $temp;
        }
    }

    function sort(array &$arr): void {
        QuickSort\sort($arr);
    }
?>

File: rust/muse/src/util/sequences.rs
use super::random_factory;

pub fn inspect(arr: &[i32]) {
    println!("[");
    arr.iter().enumerate().for_each(|(i, v)| {
        println!("\t#{:#04X}  ->  {}", i, v);
    });
    println!("]");
}

pub fn is_sorted(arr: &[i32]) -> bool {
    for i in 1..arr.len() {
        if arr[i] < arr[i - 1] {
            return false;
        }
    }

    true
}

pub fn parity_checksum(arr: &[i32]) -> i32 {
    arr.iter().fold(0, |checksum, v| checksum ^ v)
}

pub fn reverse(arr: &mut [i32]) {
    (0..arr.len() >> 1).for_each(|i| {
        let k = arr.len() - i - 1;
        arr.swap(i, k);
    });
}

pub fn shuffle(arr: &mut [i32]) {
    (0..arr.len()).for_each(|i| {
        let k = random_factory::gen_int_n(i as i32, (arr.len() - 1) as i32);
        arr.swap(i, k as usize);
    });
}

pub fn sort(arr: &mut [i32]) {
    arr.sort_unstable();
}
